### ISSUE TITLE
Dropped Messages, UDP: QoS Settings?

### ISSUE THREAD

**Issue by JeremyBYU** [continental/ecal#245]
I'm getting some dropped packets when using UDP multicast between two different computers (Sever Pub/ Client Subscribes). The specific message that keeps dropping (not received) on the client side is about 500 KB in size. The client application has many different topics that are broadcast, but only one topic  has this issue with not receiving messages. Everything works fine locally with shared memory.

I also noticed that if I pull up eca gui monitor on the client computer that it will receive the message. This means to me that the packet is probably making it through the network and onto the client computer network card buffers and so on. It might just be that the specific client **application** is subscribing to so many topics it just gets lost (PS its a python application)? I looked at `bandwidth_max_udp` for ecal.ini and it is set it to -1.  Memory size looks big enough as well (see full details below if needed. 

I was wondering if you could give some pointers on QoS settings. Is there any documentation for how to configure QoS for ecal subscriber/publisher and expected changes in behavior? Thanks!

**Desktop:**
 - OS: Ubuntu
 - Version 20.04 and 18.04
 - ECAL 5.8.5

```
; --------------------------------------------------
; NETWORK SETTINGS
; --------------------------------------------------
; network_enabled         = true / false            true  = all eCAL components communicate over network boundaries
;                                                   false = local host only communication
;
; multicast_group         = 239.0.0.1               UDP multicast group base
;
; multicast_mask          = 0.0.0.1-0.0.0.255       Mask maximum number of dynamic multicast group
;
; multicast_port          = 14000 + x               UDP multicast port number (eCAL will use at least the 2 following port
;                                                   numbers too, so please modify in steps of 10 (e.g. 1010, 1020 ...)
;
; multicast_ttl           = 0 + x                   UDP ttl value, also known as hop limit, is used in determining 
;                                                   the intermediate routers being traversed towards the destination
;
; multicast_sndbuf        = 1024 * x                UDP send buffer in bytes
;  
; multicast_rcvbuf        = 1024 * x                UDP receive buffer in bytes
;  
; bandwidth_max_udp       = -1                      UDP bandwidth limit for eCAL udp layer (-1 == unlimited)
;  
; inproc_rec_enabled      = true                    Enable to receive on eCAL inner process layer
; shm_rec_enabled         = true                    Enable to receive on eCAL shared memory layer
; udp_mc_rec_enabled      = true                    Enable to receive on eCAL udp multicast layer
;
; npcap_enabled           = false                   Enable to receive UDP traffic with the Npcap based receiver
; --------------------------------------------------

[network]
network_enabled           = true
multicast_group           = 224.0.0.1
multicast_mask            = 0.0.0.15
multicast_port            = 14000
multicast_ttl             = 4
multicast_sndbuf          = 5242880
multicast_rcvbuf          = 5242880

bandwidth_max_udp         = -1

inproc_rec_enabled        = true
shm_rec_enabled           = true
udp_mc_rec_enabled        = true

npcap_enabled             = false

; --------------------------------------------------
; COMMON SETTINGS
; --------------------------------------------------
; registration_timeout    = 60000                   Timeout for topic registration in ms (internal)
; registration_refresh    = 1000                    Topic registration refresh cylce (has to be smaller then registration timeout !)
; --------------------------------------------------
[common]
registration_timeout      = 60000
registration_refresh      = 1000

; --------------------------------------------------
; TIME SETTINGS
; --------------------------------------------------
; timesync_module_rt      = "ecaltime-localtime"    Time synchronisation interface name (dynamic library)
;                                                   The name will be extended with platform suffix (32|64), debug suffix (d) and platform extension (.dll|.so)
;
;                                                   Available modules are:
;                                                     - ecaltime-localtime    local system time without synchronization        
;                                                     - ecaltime-linuxptp     For PTP / gPTP synchronization over ethernet on Linux
;                                                                             (device configuration in ecaltime.ini)
;                                                     - ecaltime-simtime      Simulation time as published by the eCAL Player.
; --------------------------------------------------
[time]
timesync_module_rt        = "ecaltime-localtime"

; --------------------------------------------------
; ICEORYX SETTINGS
; --------------------------------------------------
; service                 = "eCAL"                  Default Iceoryx service name
; instance                = ""                      Default Iceoryx service instance name
; --------------------------------------------------

; ---------------------------------------------
; PROCESS SETTINGS
; ---------------------------------------------
;
; terminal_emulator = /usr/bin/x-terminal-emulator -e    command for starting applications with an external terminal emulator. If empty, the command will be ignored. Ignored on Windows.
;                                                            e.g.  /usr/bin/x-terminal-emulator -e
;                                                                  /usr/bin/gnome-terminal -x
;                                                                  /usr/bin/xterm -e
;
; ---------------------------------------------
[process]
terminal_emulator         = 

[iceoryx]
service                   = "eCAL"
instance                  = ""

; --------------------------------------------------
; PUBLISHER SETTINGS
; --------------------------------------------------
; use_inproc              = 0, 1, 2                 Use inner process transport layer (0 = off, 1 = on, 2 = auto, default = 0)
; use_shm                 = 0, 1, 2                 Use shared memory transport layer (0 = off, 1 = on, 2 = auto, default = 2)
; use_udp_mc              = 0, 1, 2                 Use udp multicast transport layer (0 = off, 1 = on, 2 = auto, default = 2)
;
; memfile_minsize         = x * 4096 kB             Default memory file size for new publisher
;
; memfile_reserve         = 50 .. x %               Dynamic file size reserve before recreating memory file if topic size changes
;
; memfile_ack_timeout     = 0 .. x ms               Publisher timeout for ack event from subscriber that memory file content is processed
;
; share_ttype             = 0, 1                    Share topic type via registration layer
; share_tdesc             = 0, 1                    Share topic description via registration layer (switch off to disable reflection)
; --------------------------------------------------
[publisher]
use_inproc                = 0
use_shm                   = 2
use_udp_mc                = 2

memfile_minsize           = 4096
memfile_reserve           = 50
memfile_ack_timeout       = 0

share_ttype               = 1
share_tdesc               = 1

; --------------------------------------------------
; MONITORING SETTINGS
; --------------------------------------------------
; timeout                 = 1000 + (x * 1000)       Timeout for topic monitoring in ms
; filter_excl             =  __.*                   Topics blacklist as regular expression (will not be monitored)
; filter_incl             =                         Topics whitelist as regular expression (will be monitored only)
; filter_log_con          = warning, error          Log messages logged to console (all, info, warning, error, fatal, debug1, debug2, debug3, debug4)
; filter_log_file         =                         Log messages to logged into file system
; filter_log_udp          = warning, error, fatal   Log messages logged via udp network
; --------------------------------------------------
[monitoring]
timeout                   = 5000
filter_excl               = __.*
filter_incl               =
filter_log_con            = error, fatal
filter_log_file           =
filter_log_udp            = info, warning, error, fatal

; --------------------------------------------------
; SYS SETTINGS
; --------------------------------------------------
; filter_excl             = App1,App2               Apps blacklist to be excluded when importing tasks from cloud
; --------------------------------------------------
[sys]
filter_excl               = ^eCALSysClient$|^eCALSysGUI$|^eCALSys$

```

**Comment by rex-schilasky** [comment]
Hi Jeremy, do you see the message drop counter permanently increasing in the monitor for that message ? If you can "receive" the message in the monitor, can you open it (by double clicking on it) and see the content ? How do you receive the message in your python client application, with a receive callback ? How are your machines connected, direct link or via network switch ?
No answers, just more question, sorry :-)

UDP is not a reliable protocol and eCAL is not doing any kind of QoS to keep the design simple and straight. In general messages are split in smaller packages so called MTU's (https://en.wikipedia.org/wiki/Maximum_transmission_unit). On top of this low level protocol message fragmentation, eCAL is doing it's own high level fragmentation that is dividing packages larger then 32 kb into sub messages (with message id and message counter). If ONE of such a sub message is not receiving on the client's socket (because of a missing MTU package on the lower level protocol for example) then the whole message will be dropped and the message drop counter increased by 1.

So the network adapter setting / udp configuration is playing an important role. If the problem is caused on that level you will see increasing drop counters. If your python application is to slow to process every incoming message the drop counter would not increasing but the incoming messages would overwrite not processed ones.

**Comment by JeremyBYU** [comment]
I'll have to check the drop counter in the monitor but I have answers to the other questions.

1. If I double click on the monitor on the client computer and subscribe to the to topic the message will be received and I see all the data. On the other hand the python application does not receive the data. The data clock for the topic does not increase (for python)
2. I receive with a callback.
3. Machines connected through Wifi


I think what you said last is probably what is happening: if the python process is too slow then then new messages are overwritten from non processed ones. 

As for Qos setting I just noticed that the ecal monitor said best effort and other dds implementations had settings.: https://community.rti.com/glossary/qos


Thanks for the explanation. I will probably just make this a service callback which I think uses tcp ip right?

**Comment by rex-schilasky** [comment]
Please check first if there are drops in the monitoring. 3 Hz and 500 kB is really not challenging. Can you please connect the machines via cable and recheck. Then we know that all is fine from eCAL and python processing speed.

**Comment by JeremyBYU** [comment]
Rex,

Thanks for looking into this. I just redid the tests and have a few observations. Here is a screen shot of ecal monitor to show the issue with the MeshMessage topic:
![UDP_Message](https://user-images.githubusercontent.com/6628085/114216843-70388200-9935-11eb-9848-3daaef5606ea.png)

1. The message is not dropped for the Ecal Mon process when subscribing. It is received as well and displayed (Green)
2. The Landing Server  process has a subscription to MeshMessage as well. However the data is never received on the application side. ECAl mon shows the data clock not increasing as well as no drop count increasing. (Red)

Sorry, but I dont have time to do the wired Ethernet test. I just moved this specific message over to a service instead and its seems to be reliable. **I do just want to confirm something.** **Messages sent over "services" are sent over TCP/IP correct?** I have noticed its more reliable but I have seen some "command" messages not being received sometimes over a service.  Just wanted to verify.

As for this issue we can close it because you have informed me that there are no QoS settings available for ECAL

Thanks
